nRF54L: Add new entropy driver based on the CRACEN HW

aescolar opened the pull request:
Add new entropy driver based on the CRACEN HW, and the incoming (in a PR) CRACEN CTR DRBG driver.
It also pulls new HW models to be able to use it and test it in simulation.

Note: The PR points the nRF HW models to a branch where the new HAL is forcefully enabled (ahead of its release).

aescolar committed:
modules: hal_nordic: Support new CRACEN CTR DRBG driver

Add support for the new nrf hal CRACEN CTR DRBG driver including a new
kconfig option to enable it.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
drivers: entropy: Add new driver based on nrf54l cracen driver

Add a new entropy driver based on the nrf HAL CRACEN CTR DRBG driver

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
boards nrf54l15bsim: Default to new cracen rng driver

Let's default to this new driver.
And therefore change the conditions in the BT controller kconfig
which were selecting the native_posix fake entropy driver

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
soc: nordic: nrf54L: Switch default entropy driver to new CRACEN one

Switch the default entropy driver for 54L05/10/15 devices to the new
CRACEN based one.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
manifest: Update hal_nordic with new cracen hal and rng driver

Update with the latest nordic hal

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
manifest: Update nRF hw models to latest

Update the HW models module to:
1069743a7fc06362e8f8d7f57225b8c7940da9f0

Including the following:
1069743 Fix in cracen hal conditional compile condition
3bbd3bd Enable build of hal/nrf_cracen_*.c
506da1f hal cracen_cm: Add const to prototype
b02d927 BLECrypt_if: For new aes_ecb scramble data if no libCrypto
e41f101 RRAMC nrf hal: Add replacements for new incoming RRAM APIs
4e933cf UART: Do not warn on invalid/unconfigured framesize
716b20c hal cracen_cm: Add replacements for new cracen_cm hal
fc597a4 CRACEN: Add CryptoMaster & CM AES (ECB only)
129f16c CRACEN RNG HAL: Remove senseless static inline

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
manifest: Update nRF hw models to enable cracen HAL

Update the HW models to the provisional branch zephyr_2025_01_14
which enables the use the not yet released CRACEN RNG & CM HAL

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

aescolar committed:
Bluetooth: Controller: Use cracen entropy driver for 54L devices

Use the new entropy driver for 54L devices and therefore
claim BT_CTLR_ENTROPY_SUPPORT is always supported
(note 54H remains unsuported)

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>

frkv commented on dts/bindings/rng/nordic,nrf54lcracen-ctrdrbg.yaml:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Nordic nRF54L CRACEN CTR DRBG based (Random Number Generator)
```
```suggestion
    Nordic nRF54L CRACEN CTR_DRBG based (Random Number Generator)
```
CTR_DRBG is the correct naming convention according to NIST SP800-90A

aescolar replied:
changed (in all places where it was written with a " " instead of "_")

frkv commented on drivers/entropy/Kconfig.nrf54l_cracen:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config ENTROPY_NRF54LCRACEN_CTRDRBG
+	bool "nRF54L entropy driver based on the CRACEN CTR DRBG driver"
```
```suggestion
	bool "nRF54L entropy driver based on the CRACEN CTR_DRBG driver"
```
CTR_DRBG is the correct naming convention according to NIST SP800-90A

cvinayak replied:
I guess all `CTRDRBG`  and `ctrdrbg` be `CTR_DRBG` and `ctr_drbg`, too?

aescolar replied:
AFAIK we cannot have "_" in DTS names. Anyhow, the name is long enough, so I hope it is ok without it there :)

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
LGTM

rettichschnidi commented:
At Gardena, @cmair will test this PR using an internal prototype. Since he is not (yet) part of this organization, he will give feedback himself, but I'll approve the PR in his name.

masz-nordic commented on modules/hal_nordic/nrfx/Kconfig:
```diff
@@ -25,6 +25,10 @@ config NRFX_COMP
 
+config NRFX_CRACEN
+	bool "CRACEN drivers"
+	depends on SOC_COMPATIBLE_NRF54LX
```
Should depend on DT.

aescolar replied:
There is no device tree node today for cracen itself. I could create one, but it really is just extra steps.

masz-nordic commented on dts/bindings/rng/nordic,nrf54lcracen-ctrdrbg.yaml:
```diff
@@ -0,0 +1,14 @@
+description: |
+    Nordic nRF54L CRACEN CTR_DRBG based (Random Number Generator)
+
+    This driver is only compatible with 54L devices, and may only be used from one processor
+    core. This driver cannot be used in conjunction with the nRF security PSA solution, as both
+    would attempt to use the CRACEN HW exclusively; When that is enabled, zephyr,psa-crypto-rng
+    should be selected instead.
```
Not sure if driver constraints should be included in binding description.
Perhaps it would be better to note that this node is used to describe a subset of CRACEN resources, ie. CRACEN TRNG and CryptoMaster.

aescolar replied:
> Perhaps it would be better to note that this node is used to describe a subset of CRACEN resources, ie. CRACEN TRNG and CryptoMaster.

This binding is used to describe the CTR_DRBG itself. Not the TRNG or Cryptomaster.

masz-nordic commented on dts/bindings/rng/nordic,nrf54lcracen-ctrdrbg.yaml:
```diff
@@ -0,0 +1,14 @@
+    would attempt to use the CRACEN HW exclusively; When that is enabled, zephyr,psa-crypto-rng
+    should be selected instead.
+
+compatible: "nordic,nrf54lcracen-ctrdrbg"
```
```suggestion
compatible: "nordic,cracen-ctrdrbg"
```

masz-nordic commented on drivers/entropy/entropy_nrf54l_cracen.c:
Should be renamed to `drivers/entropy/entropy_nrfx_cracen.c` and namespaces aligned.

aescolar replied:
cracen would be very ambiguous. That could be directly the TRNG or any other PRNG based on it.

masz-nordic commented on drivers/entropy/Kconfig.nrf54l_cracen:
```diff
@@ -0,0 +1,17 @@
+	bool "nRF54L entropy driver based on the CRACEN CTR_DRBG driver"
+	default y
+	depends on DT_HAS_NORDIC_NRF54LCRACEN_CTRDRBG_ENABLED
+	depends on SOC_COMPATIBLE_NRF54LX
```
```suggestion
```

aescolar commented:
Last push was just an update + nordic hal west pointer update to point to the main branch instead of a PR => no changes otherwise

cvinayak reviewed (dismissed):
subsys/bluetooth/controller updates LGTM.

anangl commented on drivers/entropy/Kconfig.nrf54l_cracen:
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2025 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config ENTROPY_NRF54LCRACEN_CTR_DRBG
```
```suggestion
config ENTROPY_NRF54L_CRACEN_CTR_DRBG
```

aescolar replied:
renames as agreed on the phone

aescolar commented:
DT, kconfig and files names changed as agreed with @anangl on a quick call.
+ other changes he requested done.

aescolar committed:
tests: bsim: cap_broadcast_reception: Workaround time dependent issue

This test is overly sensitive to the relative timing of the devices
and/or their random number generation.
Due to this the test failed with its current setup when chanding the
random generation for the nrf54.
Let's work around it by offsetting a bit one of the devices.
This is a provisional workaround until the underlying issue is
addressed.

Signed-off-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
